File: FREQNESS_Toolbox/python/src/freqness/filterFGx.py
"""Frequency-domain Gaussian filtering used by FREQ-NESS."""

from __future__ import annotations

import numpy as np
from numpy.typing import ArrayLike, NDArray
from scipy.signal import hilbert


def _nearest_index(values: NDArray[np.floating], target: float) -> int:
    return int(np.argmin(np.abs(values - target)))


def filterFGx(
    data: ArrayLike,
    srate: float,
    f: float,
    fwhm: float,
    showplot: bool = False,
) -> tuple[NDArray[np.float64], NDArray[np.float64], NDArray[np.float64]]:
    """Apply a narrow-band frequency-domain Gaussian filter.

    This is an independent Python translation of the ``filterFGx`` algorithm
    used by the FREQ-NESS MATLAB toolbox. The original filter design and
    implementations are by Mike X Cohen; see
    https://github.com/mikexcohen/GED_tutorial.

    Parameters
    ----------
    data
        Array with shape ``(channels, time)`` or a one-dimensional time series.
    srate
        Sampling rate in Hz.
    f
        Peak frequency of the filter in Hz.
    fwhm
        Full width at half maximum in Hz.
    showplot
        Plot the frequency-domain filter when ``True``.

    Returns
    -------
    filtdat
        Filtered data with the same dimensionality as the input.
    empVals
        Empirical peak frequency, frequency-domain FWHM in Hz, and temporal
        FWHM in milliseconds.
    fx
        Gain-normalized frequency-domain Gaussian.
    """
    values = np.asarray(data)
    input_was_1d = values.ndim == 1
    if input_was_1d:
        values = values[np.newaxis, :]

    if values.ndim != 2 or values.shape[1] < 2:
        raise ValueError("data must have shape (channels, time) with at least two samples")
    if not np.issubdtype(values.dtype, np.number) or np.iscomplexobj(values):
        raise TypeError("data must contain real numeric values")
    if not np.all(np.isfinite(values)):
        raise ValueError("data contains NaN or infinite values")
    if not np.isfinite(srate) or srate <= 0:
        raise ValueError("srate must be a positive finite scalar")
    if not np.isfinite(f) or f <= 0 or f >= srate / 2:
        raise ValueError("f must be positive and lower than the Nyquist frequency")
    if not np.isfinite(fwhm) or fwhm <= 0:
        raise ValueError("fwhm must be a positive finite scalar")

    n_time = values.shape[1]
    hz = np.linspace(0.0, float(srate), n_time)

    normalized_width = fwhm * (2.0 * np.pi - 1.0) / (4.0 * np.pi)
    fx = np.exp(-0.5 * ((hz - f) / normalized_width) ** 2)
    fx /= np.max(fx)

    spectrum = np.fft.fft(values, axis=1)
    filtered = 2.0 * np.real(np.fft.ifft(spectrum * fx[np.newaxis, :], axis=1))

    peak_index = _nearest_index(hz, f)
    left_index = _nearest_index(fx[: peak_index + 1], 0.5)
    right_index = peak_index + _nearest_index(fx[peak_index:], 0.5)

    impulse_envelope = np.abs(
        hilbert(np.real(np.fft.fftshift(np.fft.ifft(fx))))
    )
    impulse_envelope /= np.max(impulse_envelope)
    time = np.arange(n_time, dtype=float) / float(srate)
    temporal_peak = int(np.argmax(impulse_envelope))
    temporal_left = _nearest_index(impulse_envelope[: temporal_peak + 1], 0.5)
    temporal_right = temporal_peak + _nearest_index(
        impulse_envelope[temporal_peak:], 0.5
    )

    empirical_values = np.asarray(
        [
            hz[peak_index],
            hz[right_index] - hz[left_index],
            (time[temporal_right] - time[temporal_left]) * 1000.0,
        ],
        dtype=float,
    )

    if showplot:
        try:
            import matplotlib.pyplot as plt
        except ImportError as exc:
            raise ImportError(
                "Plotting filterFGx requires the optional 'plot' dependencies."
            ) from exc

        _, axis = plt.subplots()
        axis.plot(hz, fx, color="black", linewidth=1.3, label="Filter kernel")
        axis.set_xlim(max(f - 10.0, 0.0), f + 10.0)
        axis.set_title(
            f"Empirical filter: {empirical_values[0]:g}, "
            f"{empirical_values[1]:g} Hz"
        )
        axis.set_xlabel("Frequency (Hz)")
        axis.set_ylabel("Filter gain")
        axis.legend()

    if input_was_1d:
        filtered = filtered[0]
    return filtered, empirical_values, fx


__all__ = ["filterFGx"]

File: FREQNESS_Toolbox/FREQNESS_ExternalFunctions/filterFGx.m
function [filtdat,empVals,fx] = filterFGx(data,srate,f,fwhm,showplot)
% filterFGx   Narrow-band filter via frequency-domain Gaussian
%  [filtdat,empVals] = filterFGx(data,srate,f,fwhm,showplot)
% 
% 
%    INPUTS
%       data : 1 X time or chans X time
%      srate : sampling rate in Hz
%          f : peak frequency of filter
%       fhwm : standard deviation of filter, 
%              defined as full-width at half-maximum in Hz
%   showplot : set to true to show the frequency-domain filter shape
% 
%    OUTPUTS
%    filtdat : filtered data
%    empVals : the empirical frequency and FWHM (in Hz and in ms)
% 
% Empirical frequency and FWHM depend on the sampling rate and the
% number of time points, and may thus be slightly different from
% the requested values.
% 
% mikexcohen@gmail.com

%% input check

if nargin < 4
    help filterFGx
    error('Not enough inputs')
end

if ~isnumeric(data) || ~isreal(data) || isempty(data) || ndims(data) > 2 || ...
        any(~isfinite(data(:)))
    error('data must be a non-empty real numeric matrix containing finite values.')
end

if size(data,1) > size(data,2)
    error('data must be arranged as channels-by-time, with time along the second dimension.')
end

if size(data,2) < 2
    error('data must contain at least two time samples.')
end

if ~isnumeric(srate) || ~isscalar(srate) || ~isfinite(srate) || srate <= 0
    error('srate must be one positive finite scalar.')
end

if ~isnumeric(f) || ~isscalar(f) || ~isfinite(f) || f <= 0 || f >= srate/2
    error('f must be one positive finite scalar below the Nyquist frequency.')
end

if ~isnumeric(fwhm) || ~isscalar(fwhm) || ~isfinite(fwhm) || fwhm <= 0
    error('fwhm must be one positive finite scalar.')
end

if nargin<5
    showplot=false;
end

%% compute filter

% frequencies
ntime = size(data,2);
hz = linspace(0,srate,ntime);

% create Gaussian
s  = fwhm*(2*pi-1)/(4*pi); % normalized width
x  = hz-f;                 % shifted frequencies
fx = exp(-.5*(x/s).^2);    % gaussian
fx = fx./max(fx);          % gain-normalized

%% filter

filtdat = 2*real( ifft( bsxfun(@times,fft(data,[],2),fx) ,[],2) );

%% compute empirical frequency and standard deviation

[~,idx] = min(abs(hz-f));
empVals(1) = hz(idx);

% find values closest to .5 after MINUS before the peak
[~,idx_half_low] = min(abs(fx(1:idx)-.5));
[~,idx_half_high_rel] = min(abs(fx(idx:end)-.5));
idx_half_high = idx-1+idx_half_high_rel;
empVals(2) = hz(idx_half_high)-hz(idx_half_low);

% also temporal FWHM
tmp = abs(FREQNESS_AnalyticSignal(real(fftshift(ifft(fx)))));
tmp = tmp./max(tmp);
tx = (0:ntime-1)/srate;
[~,idxt] = max(tmp);
[~,idxt_half_low] = min(abs(tmp(1:idxt)-.5));
[~,idxt_half_high_rel] = min(abs(tmp(idxt:end)-.5));
idxt_half_high = idxt-1+idxt_half_high_rel;
empVals(3) = (tx(idxt_half_high)-tx(idxt_half_low))*1000;

%% inspect the Gaussian (turned off by default)

if showplot
    figure(10001+showplot),clf
    %subplot(211)
    plot(hz,fx,'k', 'LineWidth' , 1.3)
    hold on
    set(gca,'xlim',[max(f-10,0) f+10]);
    legend({'Filter kernel'})
    title(['Empirical filter: ' num2str(empVals(1)) ', ' num2str(empVals(2)) ' Hz' ])
    xlabel('Frequency (Hz)'), ylabel('Filter gain')
%     
%     subplot(212)
%     tmp1 = real(fftshift(ifft(fx))); tmp1 = tmp1./max(tmp1);
%     tmp2 = abs(FREQNESS_AnalyticSignal(tmp1));
%     plot(tx,tmp1, tx,tmp2), zoom on
%     xlabel('Time (s)'), ylabel('Amplitude gain')
end

%% done.
